src/Builder/FrontendDialectTransformer.cpp

```diff
@@ -1125,8 +1125,6 @@ class FrontendGenImpl {
 
   void InferTypes(const onnx::FunctionProto *func,
       std::vector<onnx::TypeProto> &inputTypes) {
-    // types: Used for temporary copies of Types, freed at end of function.
```

Comment:
from reading the code I think `types` vector solves that the lifetime of the pointer returned from node_ctx.getOutputType() cannot be counted on to extend past the lifetime of node_ctx which is only the body of the `num_inputs` for loop and ends before the pointer is read in the `typeMap` for loop afterwards

therefore I think you need to make an additional change to make it safe: change typeMap to
```
std::unordered_map<std::string, onnx::TypeProto> typeMap
```
and copy the TypeProtos into the map instead of just pointers

Reply:
btw, copying the TypeProtos into typeMap will copy the inputTypes, which is less efficient than the existing implementation

another approach, which would make everything simpler is to omit typeMap and just stick things directly into `onnx_type_map` with onnx_type_map.AddMapping instead of typeMap, but you may need to check `ContainsKey` like in 'AddValueInfo()`

Reply:
I've looked through the IR.md, onnx operators document, and the onnx.in.proto defining the base data types we're working with as part of the FrontendDialectTransformer objects. I don't have enough to see how you would rip out the node context and other bits yet, so I'm wondering if you have any more thoughts of things I should look at. You do seem to think this is possible though and I'm learning a lot which makes continued pursuit of this fix a good idea.

Reply:
I think inferTypes is improved with this change. I'm not sure there are immediate ways to improve it further but I'm open to ideas.

Otherwise there are plenty of other places in the code base to make improvements. Please keep writing PRs. If you want to run some ideas by people on the project then a good way is to open an issue to propose changes or ask questions.

Reply:
One thing I wanted to clarify is that this change doesn't really work and I would have to figure out how to factor out the nodeContext and a few other variables to really streamline this which I don't think is possible at this time and why I was hoping for a bit more feedback. If there's plenty of other places in the code though, closing this PR is probably the way to go just because I don't have enough base knowledge yet to be fully helpful.

Reply:
oops, I didn't notice that node_ctx needed the `typeMap` argument for its constructor

I think the best solution is to re-introduce the typeMap from before but make it point into onnx_type_map, i.e., all the places you call `onnx_type_map.AddMapping(name, typeproto)` you also put it into typeMap: `typeMap[name] = onnx_type_map.GetByOnnxName(name)`

Reply:
I noticed the compile error in the CI logs, looks like typeMap's 2nd type parameter needs to be a const pointer

or, if the node_ctx constructor won't accept that, you can probably work around it with a const_cast somewhere

Reply:
I double checked the assignment you saw and [maps by default will define const behind the covers](https://stackoverflow.com/questions/50593012/why-does-stdunordered-map-not-work-with-const-stdstring-key). For normal use you can just pass in directly what we need and we can move forward. I'll keep an eye out. The last few are after hours for me, so I'm not checking back as quickly.

Reply:
the stackoverflow issue is different

try `std::unordered_map<std::string, const onnx::TypeProto *> typeMap`, I think that will work because that matches 
https://github.com/onnx/onnx/blob/main/onnx/shape_inference/implementation.h#L126